dali/operators/imgcodec/image_decoder.h

```diff
@@ -0,0 +1,761 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// limitations under the License.
+
+#include <map>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+#include "dali/core/mm/memory.h"
+#include "dali/operators/decoder/cache/cached_decoder_impl.h"
+#include "dali/operators/generic/slice/slice_attr.h"
+#include "dali/operators/image/crop/crop_attr.h"
+#include "dali/operators/image/crop/random_crop_attr.h"
+#include "dali/operators/imgcodec/util/convert.h"
+#include "dali/operators/imgcodec/util/convert_gpu.h"
+#include "dali/operators/imgcodec/util/convert_utils.h"
+#include "dali/operators/imgcodec/util/nvimagecodec_types.h"
+#include "dali/operators/imgcodec/util/output_shape.h"
+#include "dali/pipeline/operator/checkpointing/stateless_operator.h"
+#include "dali/pipeline/operator/common.h"
+#include "dali/pipeline/operator/operator.h"
+
+#ifndef DALI_OPERATORS_IMGCODEC_IMAGE_DECODER_H_
+#define DALI_OPERATORS_IMGCODEC_IMAGE_DECODER_H_
+
+namespace dali {
+namespace imgcodec {
+
+template <typename Backend>
+struct OutBackend {
+  using type = GPUBackend;
+};
+
+template <>
+struct OutBackend<CPUBackend> {
+  using type = CPUBackend;
+};
+
+
+static uint32_t verbosity_to_severity(int verbose) {
+  uint32_t result = 0;
+  if (verbose >= 1)
+    result |= NVIMGCODEC_DEBUG_MESSAGE_SEVERITY_FATAL | NVIMGCODEC_DEBUG_MESSAGE_SEVERITY_ERROR;
+  if (verbose >= 2)
+    result |= NVIMGCODEC_DEBUG_MESSAGE_SEVERITY_WARNING;
+  if (verbose >= 3)
+    result |= NVIMGCODEC_DEBUG_MESSAGE_SEVERITY_INFO;
+  if (verbose >= 4)
+    result |= NVIMGCODEC_DEBUG_MESSAGE_SEVERITY_DEBUG;
+  if (verbose >= 5)
+    result |= NVIMGCODEC_DEBUG_MESSAGE_SEVERITY_TRACE;
+  return result;
+}
+
+static constexpr size_t kDevAlignment = 256;  // warp alignment for 32x64-bit
+static constexpr size_t kHostAlignment = 64;  // cache alignment
+
+static int static_dali_device_malloc(void *ctx, void **ptr, size_t size, cudaStream_t stream) {
+  auto *mr = reinterpret_cast<mm::device_async_resource *>(ctx);
+  try {
+    *ptr = mr->allocate_async(size, kDevAlignment, stream);
+    return cudaSuccess;
+  } catch (const std::bad_alloc &) {
+    *ptr = nullptr;
+    return cudaErrorMemoryAllocation;
+  } catch (const CUDAError &e) {
+    return e.is_rt_api() ? e.rt_error() : cudaErrorUnknown;
+  } catch (...) {
+    *ptr = nullptr;
+    return cudaErrorUnknown;
+  }
+}
+
+static int static_dali_device_free(void *ctx, void *ptr, size_t size, cudaStream_t stream) {
+  auto *mr = reinterpret_cast<mm::device_async_resource *>(ctx);
+  mr->deallocate_async(ptr, size, kDevAlignment, stream);
+  return cudaSuccess;
+}
+
+static int static_dali_pinned_malloc(void *ctx, void **ptr, size_t size, cudaStream_t stream) {
```

Comment:
Can the user request a "host order" allocation? Note that stream 0 is a valid stream, not "null".

Reply:
the user here are the libraries (nvjpeg, nvjpeg2k) that request allocations